### PR TITLE
PHP 8.4 | Fix passing E_USER_ERROR to trigger_error() x 3

### PR DESCRIPTION
**Pull request by jrfnl**
Notes:
* Each change is in a separate commit as each has its own reasoning and decision point.
* For the two calls changed to exceptions, it could be considered to introduce SimplePie native dedicated exceptions. This is a design choice for the maintainers of SimplePie. This PR simply fixes the PHP 8.4 compatibility issue.

---

### PHP 8.4 | Fix passing E_USER_ERROR to trigger_error() [1]

PHP 8.4 deprecates passing `E_USER_ERROR` to `trigger_error()`, with the recommendation being to replace these type of calls with either an Exception or an `exit` statement.

This commit fixes an instance found in the autoloader by changing it to a call to `exit`, which considering it is an error related to the autoloader not being registered seems the most appropriate.

Ref: https://wiki.php.net/rfc/deprecations_php_8_4#deprecate_passing_e_user_error_to_trigger_error

### PHP 8.4 | Fix passing E_USER_ERROR to trigger_error() [2]

PHP 8.4 deprecates passing `E_USER_ERROR` to `trigger_error()`, with the recommendation being to replace these type of calls with either an Exception or an `exit` statement.

This commit fixes another instance of this issue. In this case, the call to `trigger_error()` is replaced by throwing a generic `SimplePie\Exception` for the same.

Ref: https://wiki.php.net/rfc/deprecations_php_8_4#deprecate_passing_e_user_error_to_trigger_error

### PHP 8.4 | Fix passing E_USER_ERROR to trigger_error() [3]

PHP 8.4 deprecates passing `E_USER_ERROR` to `trigger_error()`, with the recommendation being to replace these type of calls with either an Exception or an `exit` statement.

This commit fixes the last instance of this issue. In this case, the call to `trigger_error()` is also replaced by throwing a generic `SimplePie\Exception` for the same.

Ref: https://wiki.php.net/rfc/deprecations_php_8_4#deprecate_passing_e_user_error_to_trigger_error

### QA: replace another call to trigger_error() with exit

... the call to `trigger_error()` would now throw a PHP notice, but with the `die()` after it, with still kill the program anyhow, so we may as well, exit with the error message and remove the `trigger_error()` function call completely.

### CODE REVIEW

**Comment by jrfnl** [comment]
Note: build failure is unrelated to this PR and should be fixed via #878

File: src/Gzdecode.php
Comment (jrfnl) [reply src/Gzdecode.php:145]: I'd be fine with that, but kept my changes to the minimum needed to fix the PHP 8.4 deprecations. Removing the method is a change which feels to me like something which should be decided by maintainer, not by a drive-by contributor like me.

File: src/SimplePie.php
Comment (jrfnl) [reply src/SimplePie.php:3299]: > `@throws` probably unnecessary

Oops, forgot to remove the `\` at the start of the name in the `throws` anyway. I'll remove the tag.

> 
> This might need to be `SimplePieException`? Or does the local `Exception` class shadow `\Exception`?

This uses the existing `SimplePie\Exception`. As the classes are in the same namespace, no import `use` statement is needed and the SimplePie `Exception` will be used.

> Or just remove it. The methods have been removed for over twelve years.

Well, yes and no. Those methods have effectively _not_ been removed, but deprecated. Removing them should probably be done in a major version and my intention with this PR is that it will be backported to the 1.8 branch for release ahead of PHP 8.4, so even though `__call()` could be removed in `master`, that wouldn't solve PHP 8.4 compatibility for 1.8.

File: src/Gzdecode.php
Comment (jrfnl) [reply src/Gzdecode.php:143]: On second thought, in this case, it's better to keep the method as it is a more stringent safeguard against dynamic properties than relying on users of this package to be running static analysis tools.

File: src/SimplePie.php
Comment (jtojnar) [reply src/SimplePie.php:3299]: > This uses the existing `SimplePie\Exception`. As the classes are in the same namespace, no import `use` statement is needed and the SimplePie `Exception` will be used.

I would expect that but then why do we have

https://github.com/simplepie/simplepie/blob/9385c54382edec0a768fdbe310c3e580c1b5dc81/src/SimplePie.php#L22

Comment (jtojnar) [reply src/SimplePie.php:3299]: > Well, yes and no. Those methods have effectively not been removed, but deprecated.

If they were just deprecated, the `__call` would proxy the call to the original methods. But they were actually both deprecated and removed and the warning was not updated.

**Review by jtojnar — APPROVED** [review]
Good enough for me. We can follow up with removal for master later.

File: src/Gzdecode.php
Comment (jtojnar) [reply src/Gzdecode.php:143]: > On second thought, in this case, it's better to keep the method as it is a more stringent safeguard against dynamic properties than relying on users of this package to be running static analysis tools.

I doubt anyone sets any properties on this, they are mostly internal state only read by the parse method and initialized at the start.

We could just as well mark them private since they are already `@access private` but that’s also for another PR.

Comment (jrfnl) [reply src/Gzdecode.php:143]: Indeed, but that still won't protect against dynamic properties (even making a class `final` won't).
Would be interesting to see what the reason was the method got added. I imagine it could be something to do with security and/or serialization ?

File: src/SimplePie.php
Comment (jrfnl) [reply src/SimplePie.php:3299]: > > This uses the existing `SimplePie\Exception`. As the classes are in the same namespace, no import `use` statement is needed and the SimplePie `Exception` will be used.
> 
> I would expect that but then why do we have
> 
> https://github.com/simplepie/simplepie/blob/9385c54382edec0a768fdbe310c3e580c1b5dc81/src/SimplePie.php#L22

Very good question, might just be to prevent confusion for contributors ?
It doesn't make a (functional) difference for this PR anyhow, though for consistency, I should maybe update the `throw new` to use the alias.

> If they were just deprecated, the __call would proxy the call to the original methods. But they were actually both deprecated and removed and the warning was not updated.

Fair enough. :+1: Sounds like a good one to clean up in 2.0 in that case (after this PR has been merged and backported to 1.8).

File: src/Gzdecode.php
Comment (jtojnar) [reply src/Gzdecode.php:143]: The class is just a pure PHP implementation of [`gzdecode`](https://www.php.net/manual/en/function.gzdecode.php) function (which was only introduced with PHP 5.4). The class is used to encapsulate the state of the stateful `gzip` decompression algorithm.

My best guess is Sam got bitten by a typo so she added it to avoid confusing debugging in the future https://github.com/simplepie/simplepie/commit/0aa49e5fbe77bd04838bf1bab7c794c095c2d5ba

We should probably deprecate it in favour of `gzdecode` too.

Comment (jrfnl) [reply src/Gzdecode.php:143]: Sounds like a good idea.

File: src/SimplePie.php
Comment (jrfnl) [reply src/SimplePie.php]: Done.

**Review by jtojnar — APPROVED** [review]

**Review by Alkarex — APPROVED** [review]

**Comment by Alkarex** [comment]
https://github.com/simplepie/simplepie/issues/884

**Review by Art4 — APPROVED** [review]

**Pull request merged**
